### DISCUSSION TITLE
Freeze layers in pytorch

### DISCUSSION THREAD

**Discussion by ghost** [lululxvi/deepxde#1160] [Q&A] [answered]
Hi, I have some quetions in using backend pytorch. Sometimes, I want to freeze some layers in net while training. But  <https://github.com/lululxvi/deepxde/blob/master/deepxde/model.py#L521-L523> and <https://github.com/lululxvi/deepxde/blob/master/deepxde/model.py#L311-L313> show that manually setting `requires_grad_(False)` is useless. After a new compile, the optimizer will continue to optimize all variables. Am I thinking right？

Thanks!

**Comment by lululxvi** [comment, marked as answer]
I am not sure how to do this in PyTorch.

**Reply by ghost** [reply]
Is it possible to remove the lines 521 and 523 in [_outputs_losses](https://github.com/lululxvi/deepxde/blob/7a53b76e94f6bf1436f0313cc63bda94cf70acc2/deepxde/model.py#L509)? so that it becomes
```python
    def _outputs_losses(self, training, inputs, targets, auxiliary_vars):
        ...
        elif backend_name == "pytorch":
            outs = outputs_losses(inputs, targets)
        ...
```
And change [trainable_variables](https://github.com/lululxvi/deepxde/blob/7a53b76e94f6bf1436f0313cc63bda94cf70acc2/deepxde/model.py#L311) to
```python
trainable_variables = (
      list(filter(lambda p: p.requires_grad, self.net.parameters())) + self.external_trainable_variables
)
```

**Reply by lululxvi** [reply]
You can test it. If it works, feel free to submit a pull request.
